### PR TITLE
More failing test fixes

### PR DESCRIPTION
**Pull request by sm6srw**
### Purpose

This Pull Request does:

* fix 17 failing tests by properly disposing the Doc Browser Extension in some Doc Browser Extension tests
* sanitize the node info doc html in order to prevent html injection attacks
* reduce false positive script tags removal log warnings
* merge latest from master

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@mjkkirschner

### CODE REVIEW

File: src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs [hunk 1]
```diff
@@ -99,7 +99,10 @@ protected virtual void Dispose(bool disposing)
             // Cleanup
             this.viewModel.LinkChanged -= NavigateToPage;
             this.documentationBrowser.Navigating -= ShouldAllowNavigation;
-            this.documentationBrowser.Dispose();
+            if (!Models.DynamoModel.IsTestMode)
+            {
+                this.documentationBrowser.Dispose();
```
Comment (mjkkirschner) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs:107]: @sm6srw can you add a comment here for future test debuggers!

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs [hunk 2]
```diff
@@ -118,24 +117,46 @@ public void Shutdown()
         protected virtual void Dispose(bool disposing)
         {
             // Cleanup
-            this.viewLoadedParamsReference.RequestOpenDocumentationLink -= HandleRequestOpenDocumentationLink;
-            this.ViewModel.MessageLogged -= OnViewModelMessageLogged;
-            documentationBrowserMenuItem.Checked -= MenuItemCheckHandler;
-            documentationBrowserMenuItem.Unchecked -= MenuItemUnCheckedHandler;
+            if (this.viewLoadedParamsReference != null)
```
Comment (mjkkirschner) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:120]: thanks for adding all this - you might want to use the new null conditional c# 6 stuff - I think you can now do:

`this.viewLoadedParamsReference?.RequestOpenDocumentationLink -= HandleRequestOpenDocumentationLink` for example.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs [hunk 3]
```diff
@@ -243,12 +247,22 @@ private string CreateNodeAnnotationContent(OpenNodeAnnotationEventArgs e)
             var writer = new StringWriter();
             try
             {
-                // Write the Node info section to the string writer
-                writer.WriteLine(NodeDocumentationHtmlGenerator.FromAnnotationEventArgs(e));
+                writer.WriteLine(DocumentationBrowserUtils.GetContentFromEmbeddedResource(STYLE_RESOURCE));
+
+                // Get the Node info section and remove script tags if any
+                var nodeDocumentation = NodeDocumentationHtmlGenerator.FromAnnotationEventArgs(e);
+                if (MarkdownHandlerInstance.RemoveScriptTagsFromString(ref nodeDocumentation))
```
Comment (mjkkirschner) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs]: I think this function has a pretty bad name now that we're using sanitizer - can we change it?

File: src/DocumentationBrowserViewExtension/MarkdownHandler.cs [hunk 4]
```diff
@@ -14,24 +14,10 @@ internal class MarkdownHandler : IDisposable
...
-                if (instance is null) { instance = new MarkdownHandler(); }
-                return instance;
-            }
-        }
-
         /// <summary>
         /// Constructor
         /// </summary>
-        private MarkdownHandler()
+        internal MarkdownHandler()
```
Comment (mjkkirschner) [inline src/DocumentationBrowserViewExtension/MarkdownHandler.cs:20]: why is this needed?

Comment (mjkkirschner) [reply src/DocumentationBrowserViewExtension/MarkdownHandler.cs:20]: ahh, to keep a public constructor from being created automatically?

File: src/Tools/Md2Html/Md2HtmlSanitizer.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using Ganss.XSS;
+
+namespace Md2Html
+{
+    /// <summary>
+    /// HtmlSanitizer class that tracks sanitizing changes
+    /// </summary>
+    internal class Md2HtmlSanitizer : HtmlSanitizer
```
Comment (mjkkirschner) [inline src/Tools/Md2Html/Md2HtmlSanitizer.cs:9]: do you expect other functionality will go here or should this have a more specific name about "changes"?

File: src/Tools/Md2Html/Md2HtmlSanitizer.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            RemovingAtRule += ChangedEvent;
+            RemovingAttribute += ChangedEvent;
+            RemovingComment += ChangedEvent;
+            RemovingCssClass += ChangedEvent;
+            RemovingStyle += ChangedEvent;
+            RemovingTag += ChangedEvent;
+        }
+
+        ~Md2HtmlSanitizer()
```
Comment (mjkkirschner) [inline src/Tools/Md2Html/Md2HtmlSanitizer.cs]: do you think this is better than disposal?

**Comment by mjkkirschner** [comment]
@sm6srw just a few comments - LGTM after that

File: src/Tools/Md2Html/Md2HtmlSanitizer.cs (hunk 6 above)
Comment (sm6srw) [reply src/Tools/Md2Html/Md2HtmlSanitizer.cs]: I am still learning about c# destructors/finalizers and IDisposal. Looks like the recommendation is to never remove event handlers in destructors as it only executes at time for garbage collect and if the object has a finalizer. Implement iDisposable is the recommended way, if needed.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs (hunk 2 above)
Comment (sm6srw) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:120]: As far as I know, the null conditional operator can not be used for assignments so it will not compile in this case.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs (hunk 3 above)
Comment (sm6srw) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs]: I agree, I will rename it.

File: src/DocumentationBrowserViewExtension/MarkdownHandler.cs (hunk 4 above)
Comment (sm6srw) [reply src/DocumentationBrowserViewExtension/MarkdownHandler.cs:20]: Correct.

File: src/Tools/Md2Html/Md2HtmlSanitizer.cs (hunk 5 above)
Comment (sm6srw) [reply src/Tools/Md2Html/Md2HtmlSanitizer.cs:9]: I don't expect any more functionality. But it is maybe clearer to talk about sanitizing updates and change the property name to ContentWasSanitized? I will do something along those lines.

**Review by mjkkirschner — APPROVED** [review]

**Comment by mjkkirschner** [comment]
@sm6srw - LGTM - let me know when you want me to merge it.

**Comment by sm6srw** [comment]
@mjkkirschner  Please merge when you have a moment.
